stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+     * @return theta Vector of solutions to the system of equations.
+     */
+    template <typename F>
+    Eigen::VectorXd
+    algebra_solver(const F& f,
+                   const Eigen::VectorXd& x,
+                   const Eigen::VectorXd& y,
+                   const std::vector<double>& dat,
+                   const std::vector<int>& dat_int,
+                   std::ostream* msgs = 0,
+                   double relative_tolerance = 1e-10,
+                   double function_tolerance = 1e-6,
+                   long int max_num_steps = 1e+3) {  // NOLINT(runtime/int)
+      // Check that arguments are valid
+      check_nonzero_size("algebra_solver", "initial guess", x);
+      check_nonzero_size("algebra_solver", "parameter vector", y);
+      for (int i = 0; i < x.size(); i++)  // FIX ME - do these w/o for loop?
+        check_finite("algebra_solver", "initial guess", x(i));
+      for (int i = 0; i < y.size(); i++)
+        check_finite("algebra_solver", "parameter vector", y(i));
+      for (size_t i = 0; i < dat.size(); i++)
+        check_finite("algebra_solver", "continuous data", dat[i]);
+
+      if (relative_tolerance <= 0)
+        invalid_argument("algebra_solver",
+                         "relative_tolerance,", relative_tolerance,
+                         "", ", must be greater than 0");
+      if (function_tolerance <= 0)
+        invalid_argument("algebra_solver",
+                         "function_tolerance,", function_tolerance,
+                         "", ", must be greater than 0");
+      if (max_num_steps <= 0)
+        invalid_argument("algebra_solver",
+                         "max_num_steps,", max_num_steps,
+                         "", ", must be greater than 0");
+
+      // Create functor for algebraic system
+      typedef system_functor<F, double, double> FS;
+      typedef hybrj_functor_solver<FS, F, double, double> FX;
+      FX fx(FS(), f, x, y, dat, dat_int, msgs, true);
+      Eigen::HybridNonLinearSolver<FX> solver(fx);
+
+      // Check dimension unknowns equals dimension of system output
+      int z_size = fx.get_value(x).size();  // FIX ME: do w/o computing fx?
+      if (z_size != x.size()) {
+        std::stringstream msg;
+        msg << ", but should have the same dimension as x "
+            << "(the vector of unknowns), which is: "
+            << x.size();
+        std::string msg_str(msg.str());
+        invalid_argument("algebra_solver", "the ouput of the algebraic system",
+                         z_size, "has dimension = ", msg_str.c_str());
+      }
+
+      // Compute theta_dbl
+      Eigen::VectorXd theta_dbl = x;
+      solver.parameters.xtol = relative_tolerance;
+      solver.parameters.maxfev = max_num_steps;
+      solver.solve(theta_dbl);
+
+      // Check if the max number of steps has been exceeded
+      if (solver.nfev >= max_num_steps)
+        invalid_argument("algebra_solver", "max number of iterations:",
```

Comment:
This doesn't seem like an invalid argument error.  It's more of a logical error.  We can throw other exception types.  

The closest thing I found was an error from Boost:

http://www.boost.org/doc/libs/1_49_0/libs/math/doc/sf_and_dist/html/math_toolkit/main_overview/error_handling.html

that's  `boost::math::evaluation_error` --- I don't know if we should be throwing those relative to the algorithms above, but @syclik or @betanalpha should know.